src/keyboard.rs

```diff
@@ -0,0 +1,1445 @@
+        self.intersects(Self::ALT)
+    }
+    /// Returns `true` if the super key is pressed.
+    pub fn super_key(&self) -> bool {
+        self.intersects(Self::SUPER)
+    }
+}
+
+bitflags! {
+    /// Represents the current state of the keyboard modifiers
+    ///
+    /// Each flag represents a modifier and is set if this modifier is active.
+    #[derive(Default)]
+    pub struct ModifiersState: u32 {
+        // left and right modifiers are currently commented out, but we should be able to support
+        // them in a future release
+        /// The "shift" key.
+        const SHIFT = 0b100;
+        // const LSHIFT = 0b010 << 0;
+        // const RSHIFT = 0b001 << 0;
+        /// The "control" key.
+        const CONTROL = 0b100 << 3;
+        // const LCTRL = 0b010 << 3;
+        // const RCTRL = 0b001 << 3;
+        /// The "alt" key.
+        const ALT = 0b100 << 6;
+        // const LALT = 0b010 << 6;
+        // const RALT = 0b001 << 6;
+        /// This is the "windows" key on PC and "command" key on Mac.
+        const SUPER = 0b100 << 9;
+        // const LSUPER  = 0b010 << 9;
+        // const RSUPER  = 0b001 << 9;
+    }
+}
+
+#[cfg(feature = "serde")]
+mod modifiers_serde {
+    use super::ModifiersState;
+    use serde::{Deserialize, Deserializer, Serialize, Serializer};
+
+    #[derive(Default, Serialize, Deserialize)]
+    #[serde(default)]
+    #[serde(rename = "ModifiersState")]
+    pub struct ModifiersStateSerialize {
+        pub shift_key: bool,
+        pub control_key: bool,
+        pub alt_key: bool,
+        pub super_key: bool,
+    }
+
+    impl Serialize for ModifiersState {
+        fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+        where
+            S: Serializer,
+        {
+            let s = ModifiersStateSerialize {
+                shift_key: self.shift_key(),
+                control_key: self.control_key(),
+                alt_key: self.alt_key(),
+                super_key: self.super_key(),
+            };
+            s.serialize(serializer)
+        }
+    }
+
+    impl<'de> Deserialize<'de> for ModifiersState {
+        fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+        where
+            D: Deserializer<'de>,
+        {
+            let ModifiersStateSerialize {
+                shift_key,
+                control_key,
+                alt_key,
+                super_key,
+            } = ModifiersStateSerialize::deserialize(deserializer)?;
+            let mut m = ModifiersState::empty();
+            m.set(ModifiersState::SHIFT, shift_key);
+            m.set(ModifiersState::CONTROL, control_key);
+            m.set(ModifiersState::ALT, alt_key);
+            m.set(ModifiersState::SUPER, super_key);
+            Ok(m)
+        }
+    }
+}
+
+/// Contains the platform-native physical key identifier (aka scancode)
+#[derive(Clone, PartialEq, Eq, Hash)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub enum NativeKeyCode {
```

Comment:
Here's a specific use case for this:
A graphical application with a settings menu where the user can select an action then press any button and the selected action will be triggered whenever the button is pressed again.

Of course, if the button is `Unidentified`, then this setting won't automatically work on other platforms, but it **should** at least work on the platform where the user made this setting. That's why we have `NativeKeyCode`, and that's why it's both serializeable and deserializeable so that the key that the user picked, can be used to trigger the action even if our API doesn't know such key.


> As for alternative, I think `Unidentified(str)` should be fine

As I explained previously, this is unfortunately not enough (but would be nice). From my other comment:
"
We want to allow applications to bind non-standard keys to actions. If such bindings are saved and synchronized across platforms, then we need to be careful, because we can't assume that a non-standard key will generate the same identifier on all platforms. That's why we also identify the platform along with the key-code.
"



> I'd suggest to split it into `NativeKeyCode` and `NativeKey`

Yeah that might be a good idea. Although that will still require having a string for browsers in each of them. (Both `key` and `code` are strings on the web. The `keyCode` is a deprecated property that is in-fact a number, but it's useless because it's 0 if the key is unidentified :D) But we can use SSO, like compact_str here as well.

Reply:
>The keyCode is a deprecated property that is in-fact a number, but it's useless because it's 0 if the key is unidentified :D) But we can use SSO, like compact_str here as well.

Then use `0`? I don't see the issue. The user will have proper virtual code on the web, and 0 as physical code. Should be fine from what I can tell, given that web is a minority platform.

>We want to allow applications to bind non-standard keys to actions. If such bindings are saved and synchronized across platforms, then we need to be careful, because we can't assume that a non-standard key will generate the same identifier on all platforms. That's why we also identify the platform along with the key-code.

I mean you can't sink such bindings at all, because they are not usable out side the system they were bind on. So what I mean is that 

Alacirtty is such application for example, and we wont use the API in that form because it's not usable. Such bindings never were portable in the first place. It just I want to have a `serde` derive that is at least usable. In the current situation it's hardly usable.

It just seems like were working around the use case that is hardly ever be used, while creating issues for serde impl.

Though, regardless of that, if you really want such thing to be used in user configs the enum must be split.

Reply:
> I mean you can't sink such bindings at all, because they are not usable out side the system they were bind on

Yeah, I agree.

> Such bindings never were portable in the first place.

Yep, exactly.

And that's precisely why we add the platform identification to the keycode, so that we avoid triggering an action for a key that happens to have the same keycode on a different platform.

> given that web is a minority platform

Alright fine, we can remove the Web variant, saying that "If the web would gain new standard keycodes that our enum doesn't have, then we should add it to our enum".

And yeah, I'm okay with splitting this into:

```rust
pub enum NativeKeyCode {
    Unidentified,
    /// Contains the extended scancode
    Windows(u16),
    /// Contains the `keyCode` from `NSEvent`
    MacOS(u16),
    XkbCode(u32),
    /// This is the android "key code" of the event as returned by
    /// `KeyEvent.getKeyCode()`
    Android(u32),
}
pub enum NativeKey {
    Unidentified,
    /// Contains the virtual-key code
    Windows(u16), 
    /// Contains the `keyCode` from `NSEvent`
    MacOS(u16),
    XkbSym(u32),
    /// This is the android "key code" of the event as returned by
    /// `KeyEvent.getKeyCode()`
    Android(u32),
}
```

Reply:
@ArturKovacs you got me slightly wrong wrt `Web`. I don't want `String` in `NativeKeyCode`, but for `NativeKey` it's fine.

Reply:
I don't have strong feelings about it. Either way works for me.

Reply:
@kchibisov I *think* the current state of the API is the way you're asking for (in your last comment, anyway).